### PR TITLE
Support T-SQL OPENQUERY

### PR DESCRIPTION
**Pull request by thephantomthief**
### Description

This commit adds a hook in the engine where we compute the tuple
descriptor from the SQL definition of a table function. If the function
is OPENQUERY, we will connect to the remote server and fetch the 
column metadata using sp_describe_first_result_set stored procedure.
Using this we modify the tuple descriptor in the analyzer and execution 
phase so that it matches with the actual result set when we run the actual
query.

Signed-off-by: Sharu Goel <goelshar@amazon.com>
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/fmgr/funcapi.c [hunk 1]
```diff
@@ -452,6 +454,18 @@ internal_get_result_type(Oid funcid,
...
 	tupdesc = build_function_result_tupdesc_t(tp);
+
+	/*
+	 * Override tupdesc for T-SQL OPENQUERY
+	 */
+	if (modify_RangeTblFunction_tupdesc_hook)
+	{
+		char* funcname = NameStr(procform->proname);
+
+		if (funcname && (strlen(funcname) == 9) && strncmp(NameStr(procform->proname), "openquery", 9) == 0)
```
Comment (kuntalghosh) [inline src/backend/utils/fmgr/funcapi.c]: This check should be moved inside the extension.

**Review by kuntalghosh — CHANGES_REQUESTED** [review]

File: src/backend/utils/fmgr/funcapi.c [hunk 2]
```diff
@@ -452,6 +454,18 @@ internal_get_result_type(Oid funcid,
...
 	tupdesc = build_function_result_tupdesc_t(tp);
+
+	/*
+	 * Override tupdesc for T-SQL OPENQUERY
+	 */
+	if (modify_RangeTblFunction_tupdesc_hook)
+	{
+		char* funcname = NameStr(procform->proname);
+
+		if (funcname && (strlen(funcname) == 9) && strncmp(funcname, "openquery", 9) == 0)
```
Comment (Deepesh125) [inline src/backend/utils/fmgr/funcapi.c]: we can save some overhead on PG by moving this operations inside hook impl. And, should it be strncasecmp?

File: src/backend/utils/fmgr/funcapi.c [hunk 3]
```diff
@@ -55,6 +55,8 @@ static bool resolve_polymorphic_tupdesc(TupleDesc tupdesc,
 										Node *call_expr);
 static TypeFuncClass get_type_func_class(Oid typid, Oid *base_typeid);
 
+modify_RangeTblFunction_tupdesc_hook_type
+	modify_RangeTblFunction_tupdesc_hook = NULL;
```
Comment (Deepesh125) [inline src/backend/utils/fmgr/funcapi.c]: Please fix indentation

File: src/backend/executor/nodeFunctionscan.c [hunk 4]
```diff
@@ -22,6 +22,7 @@
  */
 #include "postgres.h"
 
+#include "catalog/pg_proc.h"
```
Comment (Deepesh125) [inline src/backend/executor/nodeFunctionscan.c]: why do we need this?

File: src/backend/parser/parse_relation.c [hunk 5]
```diff
@@ -23,6 +23,7 @@
 #include "catalog/heap.h"
 #include "catalog/namespace.h"
 #include "catalog/pg_type.h"
+#include "executor/nodeFunctionscan.h"
```
Comment (Deepesh125) [inline src/backend/parser/parse_relation.c]: why do we need this?

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

**Review by kuntalghosh — APPROVED** [review]

**Review by Deepesh125 — APPROVED** [review]

**Pull request merged**